README.md
# mackerel-plugin-dns-lookup

This tool is a Mackerel plug-in that displays the time it takes to send a specific query to a specified DNS server for a specified number of times.There are three types of metrics that are output: "Max", "Avg", "percentile 95" "percentile 99" and "Min".

## Usage

```
Usage:
  main [OPTIONS]

Application Options:
  -s, --server=   DNS Server (default: 8.8.8.8)
  -p, --port=     query num (default: 53)
  -d, --domain=   Domain Name
  -t, --type=     Record Type
  -n, --count=    query num (default: 3)
      --timeout=  deadline time
      --protocol= Record Type (default: udp)
      --threads=  thread num (default: 1)
      --version   show version
      --debug
      --verbose

Help Options:
  -h, --help      Show this help message

Usage:
  main [OPTIONS]

Application Options:
  -s, --server=   DNS Server (default: 8.8.8.8)
  -p, --port=     query num (default: 53)
  -d, --domain=   Domain Name
  -t, --type=     Record Type
  -n, --count=    query num (default: 3)
      --timeout=  deadline time
      --protocol= Record Type (default: udp)
      --threads=  thread num (default: 1)
      --version   show version
      --debug
      --verbose

Help Options:
  -h, --help      Show this help message
```

Run (Unit is ms)

```
$ mackerel-plugin-dns-lookup \
    --server 8.8.8.8 \
    --port 53 \
    --domain google.com \
    --count 10 \
    --threads 1

min	10	1663065111
max	23	1663065111
avg	12	1663065111
p95	17	1663086632
p99	21	1663086632
```


CHANGELOG.md
# Changelog

## [v0.0.6](https://github.com/ryuichi1208/mackerel-plugin-dns-lookup/compare/v0.0.5...v0.0.6) - 2023-07-02

## [v0.0.5](https://github.com/ryuichi1208/mackerel-plugin-dns-lookup/compare/v0.0.4...v0.0.5) - 2023-07-02

## [v0.0.4](https://github.com/ryuichi1208/mackerel-plugin-dns-lookup/compare/v0.0.3...v0.0.4) - 2023-07-02

## [v0.0.3](https://github.com/ryuichi1208/mackerel-plugin-dns-lookup/compare/v0.0.2...v0.0.3) - 2023-07-02

## [v0.0.2](https://github.com/ryuichi1208/mackerel-plugin-dns-lookup/compare/v0.0.1...v0.0.2) - 2023-07-02

## [v0.0.1](https://github.com/ryuichi1208/mackerel-plugin-dns-lookup/commits/v0.0.1) - 2023-07-02


lib/dns_look_test.go
package dns

import (
	"net"
	"reflect"
	"sort"
	"sync"
	"testing"

	mp "github.com/mackerelio/go-mackerel-plugin"
)

func Test_graphGen(t *testing.T) {
	type args struct {
		labelPrefix string
		device      string
	}
	tests := []struct {
		name string
		args args
		want map[string]mp.Graphs
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := graphGen(tt.args.labelPrefix, tt.args.device); !reflect.DeepEqual(got, tt.want) {
				t.Errorf("graphGen() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestDNSResult_GraphDefinition(t *testing.T) {
	type fields struct {
		Prefix string
		Name   string
		min    float64
		max    float64
		avg    float64
		p95    float64
		p99    float64
	}
	tests := []struct {
		name   string
		fields fields
		want   map[string]mp.Graphs
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			dr := DNSResult{
				Prefix: tt.fields.Prefix,
				Name:   tt.fields.Name,
				min:    tt.fields.min,
				max:    tt.fields.max,
				avg:    tt.fields.avg,
				p95:    tt.fields.p95,
				p99:    tt.fields.p99,
			}
			if got := dr.GraphDefinition(); !reflect.DeepEqual(got, tt.want) {
				t.Errorf("DNSResult.GraphDefinition() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestDNSResult_FetchMetrics(t *testing.T) {
	type fields struct {
		Prefix string
		Name   string
		min    float64
		max    float64
		avg    float64
		p95    float64
		p99    float64
	}
	tests := []struct {
		name    string
		fields  fields
		want    map[string]float64
		wantErr bool
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			dr := DNSResult{
				Prefix: tt.fields.Prefix,
				Name:   tt.fields.Name,
				min:    tt.fields.min,
				max:    tt.fields.max,
				avg:    tt.fields.avg,
				p95:    tt.fields.p95,
				p99:    tt.fields.p99,
			}
			got, err := dr.FetchMetrics()
			if (err != nil) != tt.wantErr {
				t.Errorf("DNSResult.FetchMetrics() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("DNSResult.FetchMetrics() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestDNS_newDNS(t *testing.T) {
	type fields struct {
		r *net.Resolver
		o options
	}
	type args struct {
		o options
	}
	tests := []struct {
		name   string
		fields fields
		args   args
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			d := &DNS{
				r: tt.fields.r,
				o: tt.fields.o,
			}
			d.newDNS(tt.args.o)
		})
	}
}

func TestDNS_lookup(t *testing.T) {
	type fields struct {
		r *net.Resolver
		o options
	}
	tests := []struct {
		name    string
		fields  fields
		wantErr bool
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			d := &DNS{
				r: tt.fields.r,
				o: tt.fields.o,
			}
			if err := d.lookup(); (err != nil) != tt.wantErr {
				t.Errorf("DNS.lookup() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func Test_doQuery(t *testing.T) {
	type args struct {
		d *DNS
	}
	tests := []struct {
		name    string
		args    args
		wantRes []int64
		wantErr bool
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			gotRes, err := doQuery(tt.args.d)
			if (err != nil) != tt.wantErr {
				t.Errorf("doQuery() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !reflect.DeepEqual(gotRes, tt.wantRes) {
				t.Errorf("doQuery() = %v, want %v", gotRes, tt.wantRes)
			}
		})
	}
}

func TestDNSResult_showResult(t *testing.T) {
	type fields struct {
		Prefix string
		Name   string
		min    float64
		max    float64
		avg    float64
		p95    float64
		p99    float64
	}
	type args struct {
		res []int64
	}
	tests := []struct {
		name   string
		fields fields
		args   args
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			dr := &DNSResult{
				Prefix: tt.fields.Prefix,
				Name:   tt.fields.Name,
				min:    tt.fields.min,
				max:    tt.fields.max,
				avg:    tt.fields.avg,
				p95:    tt.fields.p95,
				p99:    tt.fields.p99,
			}
			dr.showResult(tt.args.res)
		})
	}
}

func Test_percentileN(t *testing.T) {
	type args struct {
		res []int64
		p   int
	}
	tests := []struct {
		name string
		args args
		want float64
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := percentileN(tt.args.res, tt.args.p); got != tt.want {
				t.Errorf("percentileN() = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_g(t *testing.T) {
	type args struct {
		wg *sync.WaitGroup
		d  *DNS
		c  chan<- []int64
	}
	tests := []struct {
		name    string
		args    args
		wantErr bool
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if err := g(tt.args.wg, tt.args.d, tt.args.c); (err != nil) != tt.wantErr {
				t.Errorf("g() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func Test_intToFloat64(t *testing.T) {
	type args struct {
		i []int64
	}
	tests := []struct {
		name string
		args args
		want sort.Float64Slice
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := intToFloat64(tt.args.i); !reflect.DeepEqual(got, tt.want) {
				t.Errorf("intToFloat64() = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_printPercentileN(t *testing.T) {
	type args struct {
		numbers *sort.Float64Slice
		l       int
		n       int
	}
	tests := []struct {
		name string
		args args
		want float64
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := printPercentileN(tt.args.numbers, tt.args.l, tt.args.n); got != tt.want {
				t.Errorf("printPercentileN() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestRun(t *testing.T) {
	type args struct {
		tmp []int64
	}
	tests := []struct {
		name string
		args args
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			Run(tt.args.tmp)
		})
	}
}

func TestDo(t *testing.T) {
	tests := []struct {
		name string
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			Do()
		})
	}
}


v2/client.go
package main

import (
	"bufio"
	"bytes"
	"encoding/binary"
	"fmt"
	"log"
	"net"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/jessevdk/go-flags"
)

type options struct {
	Server   string `short:"s" long:"server" description:"DNS Server" required:"false" default:"8.8.8.8"`
	Port     int    `short:"p" long:"port" description:"query num" required:"false" default:"53"`
	Domain   string `short:"d" long:"domain" description:"Domain Name" required:"true"`
	Type     string `short:"t" long:"type" description:"Record Type" required:"false"`
	Count    int    `short:"n" long:"count" description:"query num" required:"false" default:"3"`
	Timeout  int    `long:"timeout" description:"deadline time" required:"false"`
	Protocol string `long:"protocol" description:"Record Type" required:"false" default:"udp"`
	Thread   int    `long:"threads" description:"thread num" required:"false" default:"1"`
	Version  bool   `long:"version" description:"show version"`
	Debug    bool   `long:"debug" description:""`
	Verbose  bool   `long:"verbose" description:""`
}

type DNSQuery struct {
	ID           uint16
	QR           bool
	Opcode       uint8
	AA           bool
	TC           bool
	RD           bool
	RA           bool
	Z            uint8
	ResponseCode uint8
	QDCount      uint16
	ANCount      uint16
	NSCount      uint16
	ARCount      uint16
	Questions    []DNSQuestion
}

func intToFloat64(i []int64) sort.Float64Slice {
	f := make([]float64, len(i))
	for n := range i {
		f[n] = float64(i[n])
	}
	return f
}

func printPercentileN(numbers *sort.Float64Slice, l, n int) float64 {
	i := l*n/100 - 1
	ns := *numbers
	return ns[i]
}

func Run(tmp []int64) {
	dr := &DNSResult{}
	dr.showResult(tmp)

	plugin := mp.NewMackerelPlugin(dr)
	plugin.Run()

type DNSQuestion struct {
	Domain string
	Type   uint16
	Class  uint16
}

func (q DNSQuery) encode() []byte {

	q.QDCount = uint16(len(q.Questions))

	var buffer bytes.Buffer

	binary.Write(&buffer, binary.BigEndian, q.ID)

	b2i := func(b bool) int {
		if b {
			return 1
		}
		return 0
	}

	queryParams1 := byte(b2i(q.QR)<<7 | int(q.Opcode)<<3 | b2i(q.AA)<<1 | b2i(q.RD))
	queryParams2 := byte(b2i(q.RA)<<7 | int(q.Z)<<4)

	binary.Write(&buffer, binary.BigEndian, queryParams1)
	binary.Write(&buffer, binary.BigEndian, queryParams2)
	binary.Write(&buffer, binary.BigEndian, q.QDCount)
	binary.Write(&buffer, binary.BigEndian, q.ANCount)
	binary.Write(&buffer, binary.BigEndian, q.NSCount)
	binary.Write(&buffer, binary.BigEndian, q.ARCount)

	for _, question := range q.Questions {
		buffer.Write(question.encode())
	}

	return buffer.Bytes()
}

func (q DNSQuestion) encode() []byte {
	var buffer bytes.Buffer

	domainParts := strings.Split(q.Domain, ".")
	for _, part := range domainParts {
		if err := binary.Write(&buffer, binary.BigEndian, byte(len(part))); err != nil {
			log.Fatalf("Error binary.Write(..) for '%s': '%s'", part, err)
		}

		for _, c := range part {
			if err := binary.Write(&buffer, binary.BigEndian, uint8(c)); err != nil {
				log.Fatalf("Error binary.Write(..) for '%s'; '%c': '%s'", part, c, err)
			}
		}
	}

	binary.Write(&buffer, binary.BigEndian, uint8(0))
	binary.Write(&buffer, binary.BigEndian, q.Type)
	binary.Write(&buffer, binary.BigEndian, q.Class)

	return buffer.Bytes()
}

func Do(conn net.Conn, encodedQuery []byte, cnt int) (res []int64) {

	for i := 0; i < cnt; i++ {
		start := time.Now()
		conn.Write(encodedQuery)

		encodedAnswer := make([]byte, len(encodedQuery))
		if _, err := bufio.NewReader(conn).Read(encodedAnswer); err != nil {
			log.Fatal(err)
		}
		res = append(res, time.Since(start).Milliseconds())
	}

	return res
}

func Calc(wg *sync.WaitGroup, c chan<- []int64, conn net.Conn, encodedQuery []byte, cnt int) {
	defer wg.Done()
	res := Do(conn, encodedQuery, cnt)
	c <- res
}

func resultShow(tmp []int64) {
	fmt.Println(tmp)
}

func main() {
	var opts options
	_, err := flags.ParseArgs(&opts, os.Args)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	q := DNSQuestion{
		Domain: opts.Domain,
		Type:   0x1, // A record
		Class:  0x1, // Internet
	}

	query := DNSQuery{
		ID:        0xAAAA,
		RD:        true,
		Questions: []DNSQuestion{q},
	}

	// Setup a UDP connection
	conn, err := net.Dial(opts.Protocol, fmt.Sprintf("%s:%d", opts.Server, opts.Port))
	if err != nil {
		log.Fatal("failed to connect:", err)
	}
	defer conn.Close()

	if err := conn.SetDeadline(time.Now().Add(15 * time.Second)); err != nil {
		log.Fatal("failed to set deadline: ", err)
	}

	encodedQuery := query.encode()

	var wg sync.WaitGroup
	c := make(chan []int64, opts.Thread)
	for i := 0; i < opts.Thread; i++ {
		wg.Add(1)
		go Calc(&wg, c, conn, encodedQuery, opts.Count)
	}
	wg.Wait()
	close(c)

	tmp := []int64{}
	for i := range c {
		tmp = append(tmp, i...)
	}

	resultShow(tmp)

}


lib/dns_look.go
package dns

import (
	"context"
	"fmt"
	"net"
	"os"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/jessevdk/go-flags"
	mp "github.com/mackerelio/go-mackerel-plugin"
)

const Name string = "mackerel-plugin-dns-lookup"
const Version string = "0.1.0"

type options struct {
	Server   string `short:"s" long:"server" description:"DNS Server" required:"false" default:"8.8.8.8"`
	Port     int    `short:"p" long:"port" description:"query num" required:"false" default:"53"`
	Domain   string `short:"d" long:"domain" description:"Domain Name" required:"true"`
	Type     string `short:"t" long:"type" description:"Record Type" required:"false"`
	Count    int    `short:"n" long:"count" description:"query num" required:"false" default:"3"`
	Timeout  int    `long:"timeout" description:"deadline time" required:"false"`
	Protocol string `long:"protocol" description:"Record Type" required:"false" default:"udp"`
	Thread   int    `long:"threads" description:"thread num" required:"false" default:"1"`
	Version  bool   `long:"version" description:"show version"`
	Debug    bool   `long:"debug" description:""`
	Verbose  bool   `long:"verbose" description:""`
}

type DNS struct {
	r *net.Resolver
	o options
}

type DNSResult struct {
	Prefix string
	Name   string
	min    float64
	max    float64
	avg    float64
	p95    float64
	p99    float64
}

func graphGen(labelPrefix, device string) map[string]mp.Graphs {
	graphs := map[string]mp.Graphs{
		device: {
			Label: labelPrefix,
			Unit:  mp.UnitFloat,
			Metrics: []mp.Metrics{
				{Name: "min", Label: "min", Diff: false},
				{Name: "max", Label: "max", Diff: false},
				{Name: "avg", Label: "avg", Diff: false},
				{Name: "p95", Label: "p95", Diff: false},
				{Name: "p99", Label: "p99", Diff: false},
			},
		},
	}

	return graphs
}

func (dr DNSResult) GraphDefinition() map[string]mp.Graphs {
	return graphGen(fmt.Sprintf("dns-%s", strings.ReplaceAll(dr.Name, ".", "-")), "responce")
}

func (dr DNSResult) FetchMetrics() (map[string]float64, error) {
	m := make(map[string]float64)

	m["avg"] = dr.avg
	m["min"] = dr.min
	m["max"] = dr.max
	m["p95"] = dr.p95
	m["p99"] = dr.p99

	return m, nil
}

func (d *DNS) newDNS(o options) {
	d.o = o
	d.r = &net.Resolver{
		// PreferGo controls whether Go's built-in DNS resolver is preferred
		PreferGo:     true,
		StrictErrors: false,
		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
			d := net.Dialer{
				Timeout: time.Millisecond * time.Duration(o.Timeout),
			}
			return d.DialContext(ctx, o.Protocol, fmt.Sprintf("%s:%d", o.Server, o.Port))
		},
	}

}

func (d *DNS) lookup() error {
	var ip []string
	var err error
	switch d.o.Type {
	case "a", "A":
		ip, err = d.r.LookupHost(context.Background(), d.o.Domain)
	case "ptr", "PTR":
		ip, err = d.r.LookupAddr(context.Background(), d.o.Domain)
	case "txt", "TXT":
		ip, err = d.r.LookupTXT(context.Background(), d.o.Domain)
	case "cname", "CNAME":
		_, err = d.r.LookupCNAME(context.Background(), d.o.Domain)
	default:
		ip, err = d.r.LookupHost(context.Background(), d.o.Domain)
	}
	if err != nil {
		return err
	}
	if d.o.Debug && d.o.Verbose {
		for _, t := range ip {
			fmt.Printf("[DEBUG] %s\n", t)
		}
	}

	return nil
}

func doQuery(d *DNS) (res []int64, err error) {
	for i := 0; i < d.o.Count; i++ {
		start := time.Now()

		if err := d.lookup(); err != nil {
			if err != nil {
				fmt.Println(err)
			}
			return res, err
		}
		res = append(res, time.Since(start).Milliseconds())
	}

	return res, nil
}

func (dr *DNSResult) showResult(res []int64) {
	sort.Slice(res, func(i, j int) bool {
		return res[i] < res[j]
	})

	var sum int64
	for _, x := range res {
		sum += x
	}

	if len(res) > 0 {
		dr.min = float64(res[0])
		dr.max = float64(res[len(res)-1])
		dr.avg = float64(sum / int64(len(res)))
		dr.p95 = percentileN(res, 95)
		dr.p99 = percentileN(res, 99)
	}

}

func percentileN(res []int64, p int) float64 {
	i := len(res)*p/100 - 1
	if i == -1 {
		return float64(0)
	}
	return float64(res[i])
}

func g(wg *sync.WaitGroup, d *DNS, c chan<- []int64) (err error) {
	defer wg.Done()
	res, err := doQuery(d)
	if err != nil {
		return err
	}
	c <- res
	return nil
}

func intToFloat64(i []int64) sort.Float64Slice {
	f := make([]float64, len(i))
	for n := range i {
		f[n] = float64(i[n])
	}
	return f
}

func printPercentileN(numbers *sort.Float64Slice, l, n int) float64 {
	i := l*n/100 - 1
	ns := *numbers
	return ns[i]
}

func Run(tmp []int64, opts options) {
	dr := &DNSResult{
		Name: opts.Domain,
	}
	dr.showResult(tmp)

	plugin := mp.NewMackerelPlugin(dr)
	plugin.Run()
}

func Do() {
	var opts options
	_, err := flags.ParseArgs(&opts, os.Args)
	if err != nil {
		os.Exit(1)
	}

	if opts.Version {
		fmt.Fprintf(os.Stderr, "%s version %s\n", Name, Version)
		os.Exit(0)
	}

	if opts.Debug {
		fmt.Printf("[DEBUG] %v\n", opts)
		fmt.Printf("[DEBUG] QueryNum = %d\n", opts.Count*opts.Thread*2)
	}

	d := &DNS{}
	d.newDNS(opts)

	var wg sync.WaitGroup
	c := make(chan []int64, opts.Thread)
	for i := 0; i < opts.Thread; i++ {
		wg.Add(1)
		go g(&wg, d, c)
	}
	wg.Wait()
	close(c)

	tmp := []int64{}
	for i := range c {
		tmp = append(tmp, i...)
	}

	Run(tmp, opts)
}
